File: agent/lib/crypto.py
"""AES-GCM-256 field encryption with versioned wire format.

Wire format:
    version (1 byte = 0x01) || nonce (12 bytes) || ciphertext || tag (16 bytes)
"""
from __future__ import annotations

import os
from functools import lru_cache

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from agent.lib.config import Config

VERSION_BYTE = 0x01
NONCE_LEN = 12
TAG_LEN = 16
MIN_BLOB_LEN = 1 + NONCE_LEN + TAG_LEN  # version + nonce + tag (zero-length plaintext)


class CryptoError(RuntimeError):
    """Raised on decryption failure (tag mismatch, bad version, malformed blob)."""


@lru_cache(maxsize=1)
def _get_aesgcm() -> AESGCM:
    return AESGCM(Config.load().encryption_key_bytes)


def _reset_cache_for_tests() -> None:
    """Test-only: clear the cached AESGCM so a fresh ENCRYPTION_KEY takes effect."""
    _get_aesgcm.cache_clear()


def encrypt(plaintext: bytes | str) -> bytes:
    if isinstance(plaintext, str):
        plaintext = plaintext.encode("utf-8")
    aesgcm = _get_aesgcm()
    nonce = os.urandom(NONCE_LEN)
    ct_with_tag = aesgcm.encrypt(nonce, plaintext, associated_data=None)
    return bytes([VERSION_BYTE]) + nonce + ct_with_tag


def decrypt(blob: bytes) -> bytes:
    if len(blob) < MIN_BLOB_LEN:
        raise CryptoError(f"blob too short ({len(blob)} bytes; min {MIN_BLOB_LEN})")
    version = blob[0]
    if version != VERSION_BYTE:
        raise CryptoError(f"unknown crypto version byte 0x{version:02x}")
    nonce = blob[1 : 1 + NONCE_LEN]
    ct_with_tag = blob[1 + NONCE_LEN :]
    aesgcm = _get_aesgcm()
    try:
        return aesgcm.decrypt(nonce, ct_with_tag, associated_data=None)
    except Exception as exc:
        raise CryptoError("decryption failed (tag mismatch or corrupted ciphertext)") from exc

File: web/src/lib/crypto.ts
/**
 * AES-GCM-256 with a 1-byte version prefix.
 *
 * Wire format (matches agent/lib/crypto.py):
 *   [ 0x01 ][ 12-byte nonce ][ ciphertext ][ 16-byte GCM tag ]
 *
 * Key is base64-encoded 32 bytes in HIREWIRE_ENCRYPTION_KEY.
 */
import { createCipheriv, createDecipheriv, randomBytes } from 'crypto';

const VERSION = 0x01;
const NONCE_LEN = 12;
const TAG_LEN = 16;

export class CryptoError extends Error {}

function getKey(): Buffer {
  const b64 = process.env.HIREWIRE_ENCRYPTION_KEY;
  if (!b64) throw new CryptoError('HIREWIRE_ENCRYPTION_KEY not set');
  const key = Buffer.from(b64, 'base64');
  if (key.length !== 32) {
    throw new CryptoError(`HIREWIRE_ENCRYPTION_KEY must decode to 32 bytes (got ${key.length})`);
  }
  return key;
}

export function encrypt(plaintext: string): Buffer {
  const key = getKey();
  const nonce = randomBytes(NONCE_LEN);
  const cipher = createCipheriv('aes-256-gcm', key, nonce);
  const ct = Buffer.concat([cipher.update(plaintext, 'utf-8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  return Buffer.concat([Buffer.from([VERSION]), nonce, ct, tag]);
}

export function decrypt(payload: Uint8Array | Buffer): string {
  const buf = Buffer.isBuffer(payload) ? payload : Buffer.from(payload);
  if (buf.length < 1 + NONCE_LEN + TAG_LEN) {
    throw new CryptoError('ciphertext too short');
  }
  if (buf[0] !== VERSION) {
    throw new CryptoError(`unsupported wire version: 0x${buf[0].toString(16)}`);
  }
  const nonce = buf.subarray(1, 1 + NONCE_LEN);
  const tag = buf.subarray(buf.length - TAG_LEN);
  const ct = buf.subarray(1 + NONCE_LEN, buf.length - TAG_LEN);
  const key = getKey();
  const decipher = createDecipheriv('aes-256-gcm', key, nonce);
  decipher.setAuthTag(tag);
  try {
    return Buffer.concat([decipher.update(ct), decipher.final()]).toString('utf-8');
  } catch (err) {
    throw new CryptoError(`decryption failed: ${(err as Error).message}`);
  }
}
